connectors/sources/slack.py

```diff
@@ -0,0 +1,393 @@
+        )
+        while True:
+            url = f"{BASE_URL}/conversations.history?channel={channel_id}&limit={PAGE_SIZE}&oldest={oldest}&latest={latest}"
+            response = await self._get_json(url)
+            for message in response.get("messages", []):
+                latest = message.get("ts")
+                if message.get("type") == "message":
+                    if message.get("reply_count", 0) > 0:
+                        async for thread_message in self.list_thread_messages(
+                            channel_id, message["ts"]
+                        ):
+                            yield thread_message
+                    else:
+                        yield message
+            if not response.get("has_more"):
+                break
+
+    async def list_thread_messages(self, channel_id, parent_ts):
+        cursor = None
+        while True:
+            url = f"{BASE_URL}/conversations.replies?channel={channel_id}&ts={parent_ts}&limit={PAGE_SIZE}"
+            if cursor:
+                url = self._add_cursor(url, cursor)
+            response = await self._get_json(url)
+            for message in response["messages"]:
+                if message["type"] == "message":
+                    yield message
+            cursor = self._get_next_cursor(response)
+            if not cursor:
+                break
+
+    async def list_users(self, cursor=None):
+        while True:
+            url = f"{BASE_URL}/users.list?limit={PAGE_SIZE}"
+            if cursor:
+                url = self._add_cursor(url, cursor)
+            response = await self._get_json(url)
+            for member in response["members"]:
+                yield member
+            cursor = self._get_next_cursor(response)
+            if not cursor:
+                break
+
+    async def test_auth(self):
+        url = f"{BASE_URL}/auth.test"
+        response = await self._get_json(url)
+        return response.get("ok", False)
+
+    def _add_cursor(self, url, cursor):
+        return f"{url}&{CURSOR}={cursor}"
+
+    def _get_next_cursor(self, response):
+        return response.get(RESPONSE_METADATA, {}).get(NEXT_CURSOR)
+
+    async def _get_json(self, absolute_url):
+        self._logger.debug(f"Fetching url: {absolute_url}")
+        async with self._call_api(absolute_url) as resp:
+            return await resp.json()
+
+    @asynccontextmanager
+    @retryable(retries=3)
+    async def _call_api(self, absolute_url):
+        try:
+            async with self._http_session.get(
+                absolute_url,
+                headers=self._headers(),
+            ) as resp:
+                yield resp
+        except ClientResponseError as e:
+            await self._handle_client_response_error(e)
+
+    async def _handle_client_response_error(self, e):
+        if e.status == 429:
+            response_headers = e.headers or {}
+            if "Retry-After" in response_headers:
```

Comment:
Feel this can be added to `retryable`, this will be checked, and sleep if necessary, then apply the retry strategy.

Reply:
issue for DRYing up the client: https://github.com/elastic/connectors-python/issues/1304